sql-plugin/src/main/scala/org/apache/spark/sql/rapids/collectionOperations.scala

```diff
@@ -507,6 +509,136 @@ case class GpuArrayMax(child: Expression) extends GpuUnaryExpression with Implic
     }
 }
 
+case class GpuArraysZip(children: Seq[Expression]) extends GpuExpression with ExpectsInputTypes {
+
+  override def inputTypes: Seq[AbstractDataType] = Seq.fill(children.length)(ArrayType)
+
+  @transient override lazy val dataType: DataType = {
+    val fields = children.zip(arrayElementTypes).zipWithIndex.map {
+      case ((expr: NamedExpression, elementType), _) =>
+        StructField(expr.name, elementType, nullable = true)
+      case ((_, elementType), idx) =>
+        StructField(idx.toString, elementType, nullable = true)
+    }
+    ArrayType(StructType(fields), containsNull = false)
+  }
+
+  override def nullable: Boolean = children.exists(_.nullable)
+
+  @transient private lazy val arrayElementTypes =
+    children.map(_.dataType.asInstanceOf[ArrayType].elementType)
+
+  override def withNewChildrenInternal(newChildren: IndexedSeq[Expression]): Expression =
+    legacyWithNewChildren(newChildren)
+
+  override def columnarEval(batch: ColumnarBatch): Any = {
+    val numRows = batch.numRows()
+    val numCols = children.size
+    // A container collects all column vectors in use, to ease the release on exceptions.
+    val workBuffers = mutable.ArrayBuffer[ColumnVector]()
```

Comment:
I am not a fan of how workBuffers is reused all over the place. I personally would prefer to see variables that are specific to what we are doing, and what it is holding.  I would also like to see this broken down into smaller functions that are clear on what each thing is doing.

Reply:
Yes, I didn't realize I can use `safeMap` to simplify all these buffer tricks. I reorgnized all the codes: replacing workBuffers with `safeMap` and breaking the whole snippet into several functiions.